pkg/kn/commands/plugin/plugin_list_test.go

```diff
@@ -15,226 +15,208 @@
 package plugin
 
 import (
+	"bytes"
 	"fmt"
 	"io/ioutil"
 	"os"
 	"path/filepath"
+	"runtime"
 	"strings"
 	"testing"
 
 	"github.com/knative/client/pkg/kn/commands"
+	"github.com/knative/client/pkg/util"
+
 	"github.com/spf13/cobra"
 	"gotest.tools/assert"
 )
 
-func TestPluginList(t *testing.T) {
-	var (
-		rootCmd, pluginCmd, pluginListCmd      *cobra.Command
-		tmpPathDir, pluginsDir, pluginsDirFlag string
-		err                                    error
-	)
+type testContext struct {
+	pluginsDir string
+	pathDir    string
+	rootCmd    *cobra.Command
+	out        *bytes.Buffer
+	origPath   string
+}
 
-	setup := func(t *testing.T) {
-		knParams := &commands.KnParams{}
-		pluginCmd = NewPluginCommand(knParams)
-		assert.Assert(t, pluginCmd != nil)
+func (ctx *testContext) execute(args ...string) error {
+	ctx.rootCmd.SetArgs(append(args, fmt.Sprintf("--plugins-dir=%s", ctx.pluginsDir)))
+	return ctx.rootCmd.Execute()
+}
 
-		rootCmd, _, _ = commands.CreateTestKnCommand(pluginCmd, knParams)
-		assert.Assert(t, rootCmd != nil)
+func (ctx *testContext) output() string {
+	return ctx.out.String()
+}
 
-		pluginListCmd = FindSubCommand(t, pluginCmd, "list")
-		assert.Assert(t, pluginListCmd != nil)
+func (ctx *testContext) cleanup() {
+	os.RemoveAll(ctx.pluginsDir)
+	os.RemoveAll(ctx.pathDir)
+	os.Setenv("PATH", ctx.origPath)
+}
 
-		tmpPathDir, err = ioutil.TempDir("", "plugin_list")
-		assert.Assert(t, err == nil)
+func (ctx *testContext) createTestPlugin(pluginName string, fileMode os.FileMode, inPath bool) error {
+	path := ctx.pluginsDir
+	if inPath {
+		path = ctx.pathDir
+	}
+	return ctx.createTestPluginWithPath(pluginName, fileMode, path)
+}
 
-		pluginsDir = filepath.Join(tmpPathDir, "plugins")
-		pluginsDirFlag = fmt.Sprintf("--plugins-dir=%s", pluginsDir)
+func (ctx *testContext) createTestPluginWithPath(pluginName string, fileMode os.FileMode, path string) error {
+	if runtime.GOOS == "windows" {
```

Comment:
Nice. I completely forgot and ignored Windows users. But good on this.

Reply:
Though I have no way to test actual Windows plugin

Reply:
same for me ;-( I think we need at least a CI test suite somehow. I wonder how other CLI's like kubectl does it ?